File: 8_day/solve.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <string>
#include <cctype>
#include <string.h>
#include <algorithm>

using namespace std;

int getGCD(int dx, int dy) {
    if (dx < 0) dx = -dx;
    if (dy < 0) dy = -dy;
    return __gcd(dx, dy);
}

double distance(int x1, int y1, int x2, int y2) {
    return sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));
}

int main() {
    cout << "Enter file (0 for input, 1 for trial): ";
    int choice;
    cin >> choice;
    string fileName = (choice == 0) ? "input" : "trial";
    ifstream file(fileName);
    if (!file) {
        cerr << "Error: Could not open file." << endl;
        return 1;
    }

    int part1 = 0, part2 = 0;
    vector<string> lines;
    string line;
    while (getline(file, line)) {
        lines.push_back(line);
    }
    file.close();

    int m = lines.size();
    int n = lines[0].size();
    string antennas = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    bool dp[m][n];
    memset(dp, false, sizeof(dp));

    for (char antenna : antennas) {
        vector<pair<int, int>> points;

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (lines[i][j] == antenna) {
                    points.emplace_back(i, j);
                }
            }
        }

        if (points.size() == 0) continue;
        cout << antenna << ": ";
        for (auto& point : points) {
            cout << '[' << point.first << ", " << point.second << "], ";
        }
        cout << " :\n";
        
        if (points.size() == 1) continue;
        for (auto& point : points) {
            if (!dp[point.first][point.second]) {
                part2++;
                dp[point.first][point.second] = true;
            }
        }

        for (int i = 0; i < points.size(); i++) {
            for (int j = 0; j < points.size(); j++) {
                if (i == j) continue;
                int x1 = points[i].first, y1 = points[i].second;
                int x2 = points[j].first, y2 = points[j].second;

                int dx = x2 - x1;
                int dy = y2 - y1;
                int gcd = getGCD(dx, dy);
                dx /= gcd;
                dy /= gcd;
                
                int anti_node_x = x1 - dx;
                int anti_node_y = y1 - dy;
                while (anti_node_x >= 0 && anti_node_x < m && anti_node_y >= 0 && anti_node_y < n) {
                    if (!dp[anti_node_x][anti_node_y]) {
                        part2++;
                        dp[anti_node_x][anti_node_y] = true;
                    }
                    anti_node_x -= dx;
                    anti_node_y -= dy;
                }
                cout << "\n";
            }
        }
    }

    cout << "Part2: " << part1 << endl;
    return 0;
}

File: 8_day/solve.py
import string
import numpy as np
from sympy import symbols, Eq, solve, sqrt
x, y = symbols('x y')

f = open("trial", "r")
lines = [line.strip() for line in f.readlines()]
f.close()

antennas = [char for char in string.ascii_lowercase + string.ascii_uppercase + string.digits]

m: int = len(lines)
n: int = len(lines[0])
part1: int = 0

for antenna in antennas:
    points = []
    for i in range(0, m):
        for j in range(0, n):
            if lines[i][j] == antenna:
                points.append([i,j])
    if len(points) == 0: continue
    print(antenna, ": ", points)
    for i in range(0, len(points)):
        for j in range(i+1, len(points)):
            print('[', points[i], ',', points[j], ']: ', end="")
            x1,y1 = points[i]
            x2,y2 = points[j]
            for r in range(1, int(m/2)+1):
                d = np.sqrt((x2 - x1)**2 + (y2 - y1)**2)
                if r > d or d > 3*r:
                    continue
                circle1 = Eq((x - x1)**2 + (y - y1)**2, r**2)
                circle2 = Eq((x - x2)**2 + (y - y2)**2, (2*r)**2)
                linear_eq = Eq((x - x1)**2 + (y - y1)**2 - (x - x2)**2 - (y - y2)**2, r**2 - (2*r)**2)
                linear_eq = linear_eq.simplify()
                intersection_points = solve([circle1, linear_eq], (x, y))
                print('[', end="")
                for inter_point in intersection_points:
                    x_num, y_num = map(lambda p: float(p.evalf()), inter_point)
                    print('[', x_num, y_num, ']', end="")
                    if 0 <= x_num < m and 0 <= y_num < n:
                        part1 += 1
                print(']', end = " | ")
                print(part1)

print("Part1: ", part1)